ipq806x: Add support for Netgear Nighthawk Pro Gaming XR500

adamhnat opened the pull request:
Original patch from saksev from openwrt forum (https://forum.openwrt.org/t/netgear-xr500-under-dd-wrt-to-openwrt/30498/6)
Original pull request Anthoni O'Brien https://github.com/openwrt/openwrt/pull/3052 - dropped author doesn't respond for some time - I would like to continue - this pull includes code fixes according to adrianschmutzler and hnyman comments on original request
Manufacturer page: https://www.netgear.com/gaming/xr500/

Specifications:
Almost identical to Netgear R7800(https://openwrt.org/toh/netgear/r7800)
Differences:
Outside casing
Bigger flash storage
Removed esata
swapped leds: (r7800 > xr500)
usb1 (gpio 7 > 8)
usb2 (gpio 8 > 26)
guest/esata (gpio 26 > 7)

SoC: Qualcomm Atheros IPQ8065
RAM: 512 MB
Storage: 256 MiB
Wireless: Qualcomm Atheros QCA9984
Ethernet: 5x 1000/100/10
USB: 2x 3.0

Install via TFTP recovery:

1.Turn off the power, push and hold the reset button (in a hole on backside) with a pin
2.Turn on the power and wait till power led starts flashing white (after it first flashes orange for a while)
3.Release the reset button and tftp the factory img in binary mode. The power led will stop flashing if you succeeded in transferring the image, and the router reboots rather quickly with the new firmware.
4.Try to ping the router (ping 192.168.1.1). If does not respond, then tftp will not work either.

Uploading the firmware image with a TFTP client
$ tftp 192.168.1.1

    bin
    put openwrt-ipq806x-netgear_xr500-squashfs-factory.img

Signed-off-by: Adam Hnat <adamhnat@gmail.com>

adrianschmutzler commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,531 @@
+#include "qcom-ipq8065.dtsi"
```
Please add a license, same for 5.4.

adamhnat replied:
fixed, added license in 5.4 and 4.19

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,527 @@
+			drive-strength = <12>;
+			bias-disable;
+		};
+	};
```
This block can be dropped for 5.4 only, it has been moved to ipq8064.dtsi

adamhnat replied:
verified & dropped

adrianschmutzler replied:
i2c4_pins appear to be not used at all, so they can be dropped.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,527 @@
+			drive-strength = <10>;
+			bias-bus-hold;
+		};
+	};
```
This block can be dropped for 5.4 only, it has been moved to ipq8064.dtsi

adamhnat replied:
verified & dropped

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,527 @@
+			function = "mdio";
+			drive-strength = <8>;
+			bias-disable;
+		};
```
The mux block can be dropped for 5.4 only, it has been moved to ipq8064.dtsi

adamhnat replied:
verified & dropped

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,527 @@
+			function = "rgmii2";
+			drive-strength = <8>;
+			bias-disable;
+		};
```
The mux block can be dropped for 5.4 only, it has been moved to ipq8064.dtsi

adamhnat replied:
verified & dropped

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,527 @@
+			pins = "gpio18", "gpio19", "gpio21";
+			function = "gsbi5";
+			bias-pull-down;
+		};
```
Consider using the block from ipq8064.dtsi with drive-strength here as well.

adamhnat replied:
added drive-strength = <10> same as in dtsi

adrianschmutzler commented:
Please update your commit title and fix the brackets for the e-mail address in signed-off-by.

pkgadd commented:
Given that r7800 and xr500 are basically the same device, aside from the differences you noted, did you consider to merge both DTS into a common *.dtsi and only override the necessary changes in the individual *.dts files?

adamhnat commented:
applied changes according to comments. 




> Given that r7800 and xr500 are basically the same device, aside from the differences you noted, did you consider to merge both DTS into a common *.dtsi and only override the necessary changes in the individual *.dts files?

sounds like good idea, will do that on another pull reqest after finishing this one, as this should add device, and maybe other one can generalize both devices

adrianschmutzler commented:
> sounds like good idea, will do that on another pull reqest after finishing this one, as this should add device, and maybe other one can generalize both devices

I'd prefer to do this right away instead of adding the code first and then removing it again.

If it makes your life easier, note that the target is already on 5.4, so adding the DTS file for 4.19 is not strictly necessary (4.19 is just kept for testing regressions).

adrianschmutzler commented:
> I'd prefer to do this right away instead of adding the code first and then removing it again.

Since the DTS is actually almost a plain copy of the r7800, and I did my review in vain because it essentially requests what the r7800 already contains right now, I'd even more insist on unifying those right away.

adrianschmutzler commented:
I also wonder whether we can move some more stuff to qcom-ipq8065.dtsi ...
@mans0n @Ansuel Opinions?

Ansuel commented:
@adrianschmutzler what would you move from r7800 to ipq8065? 
Anyway i like the idea of the dtsi for r7800 and xr500

adrianschmutzler commented:
> @adrianschmutzler what would you move from r7800 to ipq8065?

@Ansuel That's exactly my problem: I don't know. I actually hoped you could help me here to divide into stuff that is more device-specific and stuff that's more dependent on the SoC. However, if it's not easily possible (yet), we can also just keep the current state.

Ansuel commented:
Mh I think the r7800 dts is already more or less clean with only defined router specific settings. You should check the recent changes to the ipq8064 and ipq8065 dtsi that should have removed more shared bindings.

adamhnat commented:
I've unified R7800 and XR500 into generic dtsi file as suggested, device specific settings are in .dts files - please check :-)

Ansuel commented:
Should be good

adrianschmutzler commented:
Since the labels need to contain the device name, the whole leds block needs to stay in the DTS files.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,456 @@
+
+/ {
+	model = "Netgear Nighthawk";
+	compatible = "netgear,r7800","netgear,xr500", "qcom,ipq8065", "qcom,ipq8064";
```
I'd drop compatible and model for the DTSI, it will be overwritten in DTS files anyway.

adamhnat replied:
dropped

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -1,103 +1,13 @@
-#include "qcom-ipq8065.dtsi"
+#include "qcom-ipq8065-nighthawk.dtsi"
 
 #include <dt-bindings/input/input.h>
```
This line is in the DTSI already, it can be dropped here.

adamhnat commented:
> Since the labels need to contain the device name, the whole leds block needs to stay in the DTS files.

indeed, I've also reviewed 01_leds file and there should be whole section there and aliases as well to keep aliases and led in same file

adrianschmutzler commented:
> > Since the labels need to contain the device name, the whole leds block needs to stay in the DTS files.
> 
> indeed, I've also reviewed 01_leds file and there should be whole section there and aliases as well to keep aliases and led in same file

The simplest solution to find files like that is typically to just grep for the compatible of a known similar device (in your case r7800) in target/linux/ipq806x ...

adamhnat commented:
checked and compiled with reworked dts and dtsi files - seems running fine for now

adrianschmutzler commented:
Then please squash your changes into a proper single commit again and force-push them here.

adamhnat commented:
merged into one push https://github.com/openwrt/openwrt/pull/3215/commits/8c51f56f3e70e7a6ed4d3f61389f85fb2f19b6f8

adrianschmutzler commented:
This PR still has 8 commits. Please squash with "git rebase -i ..." and then force-push the updated branch with "git push -f ...". Doing a merge typically does not help, but makes things worse.

adamhnat commented:
I had to hard reset branch and put all changes in 1 commit.... not it's final

adamhnat commented:
device is working fine for 5 days uptime, no unnoticed reboots, boot log and kernel log is also clear. I think that it's ready to commit.

adrianschmutzler commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -1,83 +1,25 @@
-#include "qcom-ipq8065.dtsi"
-
-#include <dt-bindings/input/input.h>
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Changing/adding a license to an existing file requires consent of all previous contributers.

Ansuel replied:
Should we really add a license to this? Isn't a bit overkill ?

adrianschmutzler replied:
Recently, DTS files have been required to have a license to be merged. However, this is applied to "new" files only, as contacting all people having touched an old file may be painful. So, for the case at hand, it might indeed by easiest to drop (i.e. not add) the license for this particular file.

adamhnat replied:
can we treat qcom-ipq8065-r7800.dts as "new file" as 80% if it's content was changed and were generalized into dtsi file?

adrianschmutzler replied:
No, but if you follow that path you would have to ask everybody for the DTSI as well.

adamhnat replied:
@Ansuel : as you are initial author if file https://github.com/openwrt/openwrt/commit/9a1ec4d3f4fca05275be28a6927b08c6cd43d97d#diff-7e386849ecbd95ada049a3e851054256 . Is it possible that you can consider to apply GPL-2.0-or-later OR MIT license for mentioned file?

adamhnat replied:
I've removed conflicting line in r7800 dts file so it remains without license.

adrianschmutzler commented:
Since you have removed the files-4.19 now, you should also remove the patches-4.19.

Despite, you still have formal issues with your commit title and line wrapping in commit message.

adamhnat commented:
> Since you have removed the files-4.19 now, you should also remove the patches-4.19.
> 
> Despite, you still have formal issues with your commit title and line wrapping in commit message.

The 4.19 files were almost untouched and firmware can be compiled with kernel 4.19 with dts file, so I didn't removed patches for 4.19

adrianschmutzler commented:
> > Since you have removed the files-4.19 now, you should also remove the patches-4.19.
> > Despite, you still have formal issues with your commit title and line wrapping in commit message.
> 
> The 4.19 files were almost untouched and firmware can be compiled with kernel 4.19 with dts file, so I didn't removed patches for 4.19

I thought you didn't provide a DTS file for 4.19 anymore, but appears that I've just overlooked it. Ignore my earlier comment on this subject.

adrianschmutzler commented:
Commit still has no prefix; if it gets too long, consider just using "ipq806x: add Netgear Nighthawk Pro Gaming XR500". Please remove the quotes from the commit title.

Please wrap the lines in your commit message to max. 74 characters.

adamhnat commented:
> Commit still has no prefix; if it gets too long, consider just using "ipq806x: add Netgear Nighthawk Pro Gaming XR500". Please remove the quotes from the commit title.
> 
> Please wrap the lines in your commit message to max. 74 characters.

done, according to your specification.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,417 @@
+	};
+
+	aliases {
+		serial0 = &gsbi4_serial;
```
5.4 only: serial0 can be dropped, it's in the DTSI now.

adamhnat replied:
removed from 5.4

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,417 @@
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
```
5.4 only: chosen can be dropped, it's in the DTSI now.

adamhnat replied:
also removed from 5.4

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,81 @@
+
+			reserve@e680000 {
+				label = "reserve";
+				reg = <0xe680000 0x0780000>;
```
That's only ee00000 (238 MiB) in total?

adamhnat replied:
as referenced n commit message - I've used forum patch.
Due to vendor specification there's 256MB flash. Using 238MB out of 256MB seems to be reasonable as there's UBI and remaning blocks might be used for readdress bad blocks in NAND flash.

Also checked ddwrt memory mappings and they are more strict ( https://svn.dd-wrt.com/browser/src/linux/universal/linux-4.9/arch/arm/boot/dts/qcom-ipq8065-xr500.dts )

```
reserved@e680000 {
label = "reserved";
reg = <0xe680000 0x6c0000>;
read-only;
};
```

Also I've tested commited setings for memory mapping, no any other.

adschm replied:
From partitioning I would have expected
`reg = <0xe680000 0x700000>;`
?

adschm commented:
I've cleaned this up a little and pushed it to my staging tree here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

Open questions:
- As asked above, is the last partition really ending at 0xee00000?
- Are there MAC addresses in the calibration data, i.e. at 0x1006 and 0x5006?
- You state "Removed esata", but I don't see any change in the setup corresponding to that?

adamhnat commented:
> Open questions:
> 
>     * As asked above, is the last partition really ending at 0xee00000?
I've tested those settings and worked fine for me

>     * Are there MAC addresses in the calibration data, i.e. at 0x1006 and 0x5006?
Yes, artmtd partitions contains "calibration data" including mac adress, when testing correct mac adresses were assigned to network interfaces

>     * You state "Removed esata", but I don't see any change in the setup corresponding to that?
fixed (removed unused leds)

adschm commented:
Please answer my questions in a way that allows me to update the patch in my staging tree. I certainly won't start to diff your changes against my cleanup up PR.

> Yes, artmtd partitions contains "calibration data" including mac adress, when testing correct mac adresses were assigned to network interfaces

That's two different things. I asked you whether the MAC addresses are inside calibration data (at art 0x1006 and 0x5006). In your code, you assume they are _not_ in calibration data, which is why you overwrite them with incremented addresses from 0x6. So, are there valid (and matching) MAC addresses in these locations or not? If they are, you can drop ath10k_patch_mac and will still have proper addresses.
Further reading: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

> > * As asked above, is the last partition really ending at 0xee00000?
> I've tested those settings and worked fine for me

As would have most of the other randomly chosen settings. What's the OEM's partitioning on this device?

> > * You state "Removed esata", but I don't see any change in the setup corresponding to that?
> fixed (removed unused leds)

So, what do I have to change?

adamhnat commented:
I was going to check yesterday flash data for calbration, and put here partition layout from vendor... however my router is bricked now (were running fine, then I came back home and it's dead (no serial response - I will try to repair it, as it's out of warranty)

also regarding SATA port - it was phisically removed (on board there still place for it) and also LED was removed - I've removed sata led from dts for xr500 config.

unfortunatly until I repair my device I cannot help more. My crash was on vendor firmware (not openwrt related issue, more hardware)

JordanP-Dev commented:
Hi @adschm.  I built OpenWRT using the commit adding XR500 in your staging repo and installed it.  Does that affect the art partition at all?

If not, when looking at /dev/mtd3, I see 3 mac addresses at 0x0, 0x6, and 0xc.  I think 0x0 should be assigned to the WAN, as it matches the MAC address on the bottom of the router.  It currently gets assigned to the LAN.

adschm commented:
> Hi @adschm. I built OpenWRT using the commit adding XR500 in your staging repo and installed it. Does that affect the art partition at all?
> 
> If not, when looking at /dev/mtd3, I see 3 mac addresses at 0x0, 0x6, and 0xc. I think 0x0 should be assigned to the WAN, as it matches the MAC address on the bottom of the router. It currently gets assigned to the LAN.

So, there are no valid-looking addresses in 0x1002 or 0x5006?

The MAC address on the label isn't necessarily linked to a specific interface. The addresses should essentially be used the same way as in OEM firmware, if possible.

JordanP-Dev commented:
> So, there are no valid-looking addresses in 0x1002 or 0x5006?

No, there are not.
 
> The MAC address on the label isn't necessarily linked to a specific interface. The addresses should essentially be used the same way as in OEM firmware, if possible.

I'll stick the OEM firmware this weekend and have a look.

adschm commented:
> > The MAC address on the label isn't necessarily linked to a specific interface. The addresses should essentially be used the same way as in OEM firmware, if possible.
> 
> I'll stick the OEM firmware back on this weekend and have a look.

That would be nice. Essentially, we just need the four addresses (or rather, just the last byte) for the four relevant interfaces: lan, wan, 2.4g, 5g. If we know that we just pick the wrong addresses from flash for OpenWrt then.

Which addresses are in 0x0, 0x6, 0xc for your device? (Last byte is enough)

JordanP-Dev commented:
> Which addresses are in 0x0, 0x6, 0xc for your device? (Last byte is enough)

ee, ef, and f0

JordanP-Dev commented:
On the OEM firmware, the mac addresses are:

WAN: ef
LAN: ee
2G: ee
5G: f0

The partition layout from /proc/mtd is:

dev:    size   erasesize  name
mtd0: 00c80000 00020000 "qcadata"
mtd1: 00500000 00020000 "APPSBL"
mtd2: 00080000 00020000 "APPSBLENV"
mtd3: 00140000 00020000 "ART"
mtd4: 00140000 00020000 "ART.bak"
mtd5: 00220000 00020000 "kernel"
mtd6: 02fe0000 00020000 "rootfs"
mtd7: 0a000000 00020000 "netgear"
mtd8: 03200000 00020000 "firmware"
mtd9: 00080000 00020000 "crashdump"
mtd10: 00380000 00020000 "language"
mtd11: 00120000 00020000 "config"
mtd12: 00120000 00020000 "pot"
mtd13: 00080000 00020000 "cert"
mtd14: 00010000 00001000 "m25p80"
mtd15: 0005d000 0001f000 "pot.bak"
mtd16: 001b2000 0001f000 "traffic_meter"
mtd17: 001b2000 0001f000 "traffic_meter.bak"
mtd18: 001b2000 0001f000 "dongle"
mtd19: 037b4000 0001f000 "overlay_volume"

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,412 @@
+			reg = <0x00010000 0 0 0 0>;
+
+			mtd-mac-address = <&art 6>;
+			mtd-mac-address-increment = <(1)>;
```
Which one is 2g and which 5g here? Based on the recent MAC address findings, 2g should get 0x0 and 5g should get 0xc directly.

Apart from that, MAC addresses are initialized twice, once here in DTS and once in target/linux/ipq806x/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
One init should be enough.

Ansuel replied:
think the original idea was remove the hotplug script make the dts do all the work...  (this is present only on r7800 and xr500... but should be done also on all the other target)

adschm replied:
So depending on what is in the caldata at the MAC address position the caldata will overwrite DTS or not ...?

adschm replied:
@JordanP-Dev So maybe you can build an OpenWrt image and remove the two "ath10k_patch_mac" lines in 11-ath10k-foo file, and then report which address 2g/5g get on the device?

Ansuel replied:
@adschm ok i found it.... I didn't check but the mtd-mac-address is not supported at all by ath10k...
This was introduced by chuckeey and support was added with this testing patch
https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=4747c97c090a161b8d4240c1acd03502316d9269

I'm checking if these patch actually got merged...

Ansuel replied:
Ok these patch got merged and are present in our ath10k-ct

JordanP-Dev commented:
Unfortunately, I think my XR500 is bricked now.  After I reinstalled the OEM firmware with tftp, it started up fine.  After powering it off, it won't boot back up.  When I turn it on, the USB2 light comes on white and stays white, while the eth lights flash amber, then turn off.  If there is an ethernet cable plugged in, that eth light will turn on.  I don't seem to be able to get it into tftp mode.

@adamhnat.  Is this what happened to yours?

Ansuel commented:
a serial output would be very helpful probably the stock firmware do some funny thing to uboot... Anyway try to plug the eth cable before powering it on and do the recovery procedure with the reset button

Ansuel commented:
Anyway this should always work since it does run on every boot (even without the reset button pressed)
https://github.com/jclehner/nmrpflash

JordanP-Dev commented:
I've ordered a serial cable since I don't have one on hand.

It will not go into tftp mode at all.  nmrpflash does not get any communication from the router when run.

JordanP-Dev commented:
Weird.  It seems to be acting as a plain switch.  If I connect a computer to the XR500, and connect the XR500 to my other router, the computer will get a dhcp lease from my other router.

JordanP-Dev commented:
Got my usb serial adapter today.  There is no communication from the router at all.  I've found somebody else who had the same issue, but no resolution:

https://forum.netduma.com/topic/30466-xr500-wifi-not-showing/

There also seems to be something similar that can happen to the r7xxx series, where the esata lights up instead.  I've found several examples.

https://community.netgear.com/t5/Nighthawk-WiFi-Routers/Bricked-No-power-LED-ethernet-LEDs-blinking/td-p/504434

https://community.netgear.com/t5/Nighthawk-WiFi-Routers/My-nighthawk-only-displays-a-solid-white-light-under-eSATA-It/td-p/1724471

https://community.netgear.com/t5/Nighthawk-WiFi-Routers/R7800-shut-down-unexpectedly-and-now-only-eSATA-led-will-light/td-p/1708809

https://community.netgear.com/t5/DSL-Modems-Routers/D7800-zero-connectivity/td-p/1448295

https://www.snbforums.com/threads/need-help-bricked-my-r7800.37054/

https://community.netgear.com/t5/Nighthawk-WiFi-Routers/Nighthawk-x4-r7500-on-showing-esata-led/td-p/1190209

https://community.netgear.com/t5/Nighthawk-WiFi-Routers/Dreadful-Netgear-s-support/td-p/1180758

Ansuel commented:
Are you sure you connected the right way? (tx to rx on the usb serial... rx to tx and gnd without connecting the vcc)

JordanP-Dev commented:
I looked at the [r7800 page](https://openwrt.org/toh/netgear/r7800) for reference because I assumed it should be the same.  I tried to double check using the instructions here:

https://openwrt.org/docs/techref/hardware/port.serial#voltmeter_scheme

The r7800 article labels the pins based on the USB-ttl markings, so I think the pins from the outside-in should be GND, TX, RX, 3.3V.  If that is correct, I get the following:

3.3v from VCC -> GND.
0v from VCC -> RX
0v from VCC -> TX
0v from RX -> GND
0.1v from TX -> GND

This doesn't match the voltmeter scheme listing for RX -> GND and TX -> GND, so I'm not really sure.

I've hooked up my adapter with TX and RX hooked up both ways, and both receive nothing.  I've done a loopback test on my adapter, and that works properly.

JordanP-Dev commented:
> I mean it's too strange that things broke up with no excessive stress (i don't think you flashed a lot of image on that)

I only flashed it twice.  First flash was to try the OpenWRT build.  The second flash was to install the latest XR500 firmware.  After installing, it loaded the firmware fine.  After powering it off and back on it did this.

adamhnat commented:
> @adamhnat. Is this what happened to yours?
Exactly same thing, seems that there are some more flash layout issues that put device on this state. Also before bricing my srial consile was working fine, after it - nothing is on serial lines. Router is working in "switch mode".

Ansuel commented:
A muted serial looks like a wiped uboot (no bootloader)

Wonder if there is a bug in the stock firmware that misscalculate the partition layout and wrongly write on the bootloader section instead of somewhere else... Could be caused by the fact that we remove the netgear partition and there is some script in the stock firmware that read some offset in that section, the offset in that section is 0 (due to the fact that we wiped it) and the script write data at the start of the nand... The first boot works... You reboot and you bricked the router...

JordanP-Dev commented:
The latest stock firmware is:

http://www.downloads.netgear.com/files/GDC/XR500/XR500-V2.3.2.56.zip

I also have raw dumps of the mtd partitions from when OpenWRT was installed, if that would be useful.

adamhnat commented:
also there's source of older firmware version: https://www.downloads.netgear.com/files/GPL/XR450-XR500-V2.3.2.40_gpl_src.tar.bz2.zip     (to compile it you need ubuntu 12.10 )

Ansuel commented:
With a quick search and comparison with r7800 i can't find anything that evident that would cause problems... I still suspect there is some magic with flash_erase or dd or mtd but the uboot partition should be readonly AFAIK.
Anyway no way to use the warranty?

adamhnat commented:
> Anyway no way to use the warranty?

In my case that's not possible.. At begging I had device that wasn't powering on (after service market, no antennas, no ac adapter), I resolved powered on issue (12V -> 5V converted was broken) then device powered up without any issue, and run fine for few months... until I started to experiments. 
I also though that uboot should be read only and in worst case I will have serial console - I was wrong.

JordanP-Dev commented:
I don't think they will replace it if it has been modified, will they?

Ansuel commented:
Well... software modification are not really modification... They also have a recovery mode to restore things... To me this looks like a device defect and they should replace the device. (it's their fault if they have a firmware that can destroy the device... also it's the stock firmware that destroyed it not the custom one)

adamhnat commented:
I don't have even an option (my warranty info:)
Product Warranty Information

```
Product: XR500
Serial Number: XXXXXXXXXXXXXXXX
Country of Purchase: Poland
Purchased on: 2019-02-09

Standard Product Warranty:
Hardware Warranty: Expired
Phone Support: Expired
Email Support: Expired
Chat Support: Expired
Power Supply Warranty: Expired
Accessories Warranty: Expired
```

I hope that there will be a day that some hardware magican will find out how to recover uboot on ipq806x devices / there will be some leak from devs that could help restore it using JTAG and openocd/any other method.

pgwipeout commented:
@adamhnat If the bootloader is corrupt or missing, qualcomm devices enumerate as a usb recovery device.
Have you tried plugging in with a usb a to usb a cable to a computer?

adamhnat commented:
> @adamhnat If the bootloader is corrupt or missing, qualcomm devices enumerate as a usb recovery device.
> Have you tried plugging in with a usb a to usb a cable to a computer?

Yes, that's is my second guess, but unfortunatly no usb devices detected - maybe I'm doing something wrong with wiring? I've connected gnd-gnd, d+ -> d-, d- -> d+, vcc left disconnected. I also tryed to short d- and gnd in boot, but no success at all.

pgwipeout commented:
No, Data+ goes to Data+, Data- goes to Data-, Gnd goes to Gnd, and VCC +5 from the host is how it knows a host is connected, so that needs to be connected as well.

adamhnat commented:
> 
> 
> No, Data+ goes to Data+, Data- goes to Data-, Gnd goes to Gnd, and VCC +5 from the host is how it knows a host is connected, so that needs to be connected as well.

Still no luck, resolidered cables as you suggested  - still no luck - no usb device appears, on both USBs - seems that this EDL mode is valid for APQ and MPQ series of processors.

pgwipeout commented:
> Still no luck, resolidered cables as you suggested - still no luck - no usb device appears, on both USBs - seems that this EDL mode is valid for APQ and MPQ series of processors.

It might be that the bootrom is finding something valid to boot, but when it jumps there it's broken.
You might have to force the mode manually.

I'm currently disassembling the firmware and sources for stock trying to find hints as to what is wrong.
I've found indications the partitions might be wrong, but nothing concrete.
It also seems to be missing a replacement u-boot in all of the firmwares I've looked at so far.

pgwipeout commented:
Some more info:
You have to short D+ not D- to ground in order to force EDL mode.
Another method is by shorting JTAG CMD pin to ground.

Looks like there is a JTAG header on board.

pgwipeout commented:
> On the OEM firmware, the mac addresses are:
> 
> WAN: ef
> LAN: ee
> 2G: ee
> 5G: f0
> 
> The partition layout from /proc/mtd is:
> 
> dev: size erasesize name
> mtd0: 00c80000 00020000 "qcadata"
> mtd1: 00500000 00020000 "APPSBL"
> mtd2: 00080000 00020000 "APPSBLENV"
> mtd3: 00140000 00020000 "ART"
> mtd4: 00140000 00020000 "ART.bak"
> mtd5: 00220000 00020000 "kernel"
> mtd6: 02fe0000 00020000 "rootfs"
> mtd7: 0a000000 00020000 "netgear"
> mtd8: 03200000 00020000 "firmware"
> mtd9: 00080000 00020000 "crashdump"
> mtd10: 00380000 00020000 "language"
> mtd11: 00120000 00020000 "config"
> mtd12: 00120000 00020000 "pot"
> mtd13: 00080000 00020000 "cert"
> mtd14: 00010000 00001000 "m25p80"
> mtd15: 0005d000 0001f000 "pot.bak"
> mtd16: 001b2000 0001f000 "traffic_meter"
> mtd17: 001b2000 0001f000 "traffic_meter.bak"
> mtd18: 001b2000 0001f000 "dongle"
> mtd19: 037b4000 0001f000 "overlay_volume"

Okay there's an error with kernel and ubi partitions.
You missed the kernel partition changed sizes which warps everything after it.
On the r7800 kernel is at `reg = <0x1480000 0x0400000>;`
On the xr500, kernel is at `reg = <0x1480000 0x0220000>;`
This means ubi on the xr500 (labeled rootfs) is at `reg = <0x16a0000 0x2fe0000>;`

Honestly we should also have all of the other partitions labeled as well, or at least everything past 0x4680000 marked as read only.

I still haven't found where it expects to put u-boot yet though.

Ansuel commented:
@pgwipeout Wonder how this caused the brick of the router (anything after the kernel is just data that gets recreated by the original firmware AFAIK)

pgwipeout commented:
> @pgwipeout Wonder how this caused the brick of the router (anything after the kernel is just data that gets recreated by the original firmware AFAIK)

It depends on if the partition tables are hard coded in certain places.
I found a reference to updating uboot in the original firmware, but it's an empty file.
Both reports say it booted once into the netgear firmware then stopped booting altogether.
I wonder if once we boot into linux it's running a u-boot update but updating it with nothing.

> 
> Wonder if the uboot fw_recovery is bugged... (overflows and start to write to the start of the nand?) But this would brick it from the the next reboot... It's definitely something from the netgear firmware

I'm building the original firmware from netgear's source to see the final products before they are packaged and to compare to the official firmware blobs.
I've also posted all versions of the xr500 source to a git repo.
https://github.com/pgwipeout/xr500-firmware 
It will build on ubuntu 14, but it requires a change to account for a build script issue.

Ansuel commented:
Take account that the original firmware can be easily extracted using binwalk.

pgwipeout commented:
> Take account that the original firmware can be easily extracted using binwalk.

Indeed, that's what I'm doing as well.
Unfortunately manufacturers tend to *fib* on their source releases, so comparing the two can be useful.
It's also nice to see all of the products of the build, even if the finished file is the same.

pgwipeout commented:
For instance I'm curious if u-boot is being built, and simply not packaged in the official netgear release.

Ansuel commented:
@pgwipeout From what i know, they just provide uboot source (if you are lucky with the version) but it's not actually used...

pgwipeout commented:
Is there a way on the official firmware to get a root shell without a serial port?
I'm on vacation right now and don't have a ttl converter with me.
I'd like to dump the entire NAND from a fresh board, install openwrt, revert back to stock, then dump the flash again to see what has changed.

Ansuel commented:
> @Ansuel Thank you!
> @adamhnat @JordanP-Dev Which versions of firmware were your devices on originally, and which did you try flashing?
> I found a reference to someone who did a web upgrade to XR500-V2.0.0.28 and it soft bricked.
> https://community.netgear.com/t5/Nighthawk-Pro-Gaming-Routers/XR500-Firmware-got-corrupted-after-upgrade-Brick-Router-HELP/td-p/1822087
> They tried tftp2 but it failed, but tftpd64 worked.

think they take the router and directly flash openwrt... So the original firmware was installed?

TurboFerris commented:
@pgwipeout I flashed openwrt over the original firmware.  Then I flashed XR500-V2.3.2.56.zip.

JordanP-Dev commented:
@pgwipeout I flashed openwrt over the original firmware.  Then I flashed XR500-V2.3.2.56.zip.

adamhnat commented:
@pgwipeout : my device was bricked on version 2.3.2.56 
Regarding this NG topic - it's softbrick - uboot is alive as power led is orange. In out case when push power button in back of router only half leds blinks (normally all of them blink once), then only USB2 led is white, when connect network cable into ethernet, also coresponding led starting be white.

I've tryed as you sugested short d+ with gnd - but still no luck-no usb devices detected.  by "JTAG CMD" did you mean TDI or TDS Jtag pin?

If you need raw uboot - I've manage to compile firmware and uboot as separate binary from GPL sources version 2.3.2.40

adamhnat commented:
regarding uboot place in sources there's documentation how to upgrade uboot:
(IPQ) # tftpboot 0x42000000 u-boot.mbn
(IPQ) # crc32 0x42000000 ${filesize}
(IPQ) # ipq_nand sbl && nand erase 0x00c80000 0x00580000 && nand write 0x42000000 0x00c80000 0x80000
(IPQ) # reset

Ansuel commented:
This is the same for every uboot?

adamhnat commented:
From sources there's also documentation for other device, here are other examples of other netgear devices (didn't test it on live device):
for **R7800**:

(IPQ) # tftpboot 0x42000000 u-boot.mbn
(IPQ) # crc32 0x42000000 ${filesize}
(IPQ) # ipq_nand sbl && nand erase 0x00c80000 0x00580000 && nand write 0x42000000 0x00c80000 0x80000
(IPQ) # reset

Steps to upgrade u-boot through single image:
=============================================

(IPQ) # setenv machid 136c
(IPQ) # tftpboot 0x42000000 u-boot-single.img
(IPQ) # crc32 0x42000000 ${filesize}
(IPQ) # ipq_nand linux && nand erase 0x01180000 0x00080000 && imgaddr=0x42000000 && source $imgaddr:script
(IPQ) # reset


**FOR R7500v2:**
(IPQ) # tftpboot 0x42000000 u-boot.mbn
(IPQ) # crc32 0x42000000 ${filesize}
(IPQ) # ipq_nand sbl && nand erase 0x00c80000 0x00580000 && nand write 0x42000000 0x00c80000 0x80000

Steps to upgrade u-boot through single image:
=============================================
(IPQ) # tftpboot 0x42000000 u-boot-single.img
(IPQ) # crc32 0x42000000 ${filesize}
(IPQ) # ipq_nand sbl && nand erase 0x00c80000 0x00580000 && imgaddr=0x42000000 && source $imgaddr:script

adamhnat commented:
> Is there a way on the official firmware to get a root shell without a serial port?
> I'm on vacation right now and don't have a ttl converter with me.
> I'd like to dump the entire NAND from a fresh board, install openwrt, revert back to stock, then dump the flash again to see what has changed.

yes, on first boot enter http://192.168.1.1/debug.htm and enable telnet checkbox. after that telnet to router.

pgwipeout commented:
Okay, so APPSBL is u-boot, which means APPSBLENV is likely the u-boot environment.

I've looked at the USB configuration and both ports are using designware xhci IP. 
So they are capable of doing it, but it is up to the bootrom if it's used.
I found this email chain for OpenOCD and the ipq806x if you have a jtag interface 
https://www.mail-archive.com/openocd-devel@lists.sourceforge.net/msg08770.html
The TLDR of it, he ended up just reprogramming the nand chip manually.

adamhnat commented:
> Okay, so APPSBL is u-boot, which means APPSBLENV is likely the u-boot environment.
> 
> I've looked at the USB configuration and both ports are using designware xhci IP.
> So they are capable of doing it, but it is up to the bootrom if it's used.
> I found this email chain for OpenOCD and the ipq806x if you have a jtag interface
> https://www.mail-archive.com/openocd-devel@lists.sourceforge.net/msg08770.html
> The TLDR of it, he ended up just reprogramming the nand chip manually.

I also found that - however in this device NAND is VFBGA type, not TSOP - so there's no easy way to reprogram it. I have jtag interface (but only 5V and 3.3V) so I ordered logic converter, as I found that JTAG is 1.8V also seems like it's standard JTAG 20P for ARM. 

My another finding is: http://www.hackspot.net/eeroBlog   - he manage to find SPI pins for memory + CPI SPI bus release pin to reprogram nand.

adamhnat commented:
> > @pgwipeout Wonder how this caused the brick of the router (anything after the kernel is just data that gets recreated by the original firmware AFAIK)
> 
> It depends on if the partition tables are hard coded in certain places.
> I found a reference to updating uboot in the original firmware, but it's an empty file.
> Both reports say it booted once into the netgear firmware then stopped booting altogether.
> I wonder if once we boot into linux it's running a u-boot update but updating it with nothing.
> 
> > Wonder if the uboot fw_recovery is bugged... (overflows and start to write to the start of the nand?) But this would brick it from the the next reboot... It's definitely something from the netgear firmware
> 
> I'm building the original firmware from netgear's source to see the final products before they are packaged and to compare to the official firmware blobs.
> I've also posted all versions of the xr500 source to a git repo.
> https://github.com/pgwipeout/xr500-firmware
> It will build on ubuntu 14, but it requires a change to account for a build script issue.

I've tryed also to build it on another OS than ubuntu 12.10 x86 - it won't work (too many error and deps to resolve including ulibc and gcc version). I manage to create VM with this OS and successful build firmware and uboot as well.

pgwipeout commented:
So I managed to restore from OpenWRT to XR500-V2.0.0.28.img without a hard brick.
My build of OpenWRT didn't come up on the network for some reason.

pgwipeout commented:
> I've found that official firmware for version 2.3.2.40 is 33MB and build from GPL sources is 25MB - so there must be change.

Probably the precompiled kernel modules included in the official version.

pgwipeout commented:
So it would seem mtd0 and mtd1 are protected from reading in linux userspace.

pgwipeout commented:
Okay, I've got it working and it works well.
So Tested-By: Peter Geis <pgwipeout@gmail.com>

Note, recovery to XR500-V2.0.0.28.img is recommend vice the newest firmware until the hard brick issue is addressed.

Ansuel commented:
@pgwipeout you reported wrong partition declaration... can you point how they should be fixed?

pgwipeout commented:
@Ansuel It doesn't match the original, but it is functional.
Current partition table increases the kernel partition size and significantly decreases the ubi partition size.
Both partitions still reside within the combined space of kernel/ubi. 
Additionally the overlay partition from the original kernel is untouched, so returning to stock retains stock settings.

I still recommend that all partitions be mapped as they are originally, and all partitions not touched should be marked read only.
But it does work as it is.

Ansuel commented:
On r7800 we safely overwrite the netgear partition (since the firmware regens it) 
How much space we have in the overlay with the current table?

pgwipeout commented:
Currently we have 206M for mtd6.

Ansuel commented:
Oh ok... 206 are enough ahahha

adamhnat commented:
small update: I measured power before coil (12.4V) and after coli - 1.08V - seems that I have power issue not related with firmware. I found out that wifi and wps led are lighting very poor.  Some element accidentally has broken - I have hope that's not SoC - I will continue investigation.

pgwipeout commented on target/linux/ipq806x/patches-4.19/0069-arm-boot-add-dts-files.patch:
```diff
@@ -10,7 +10,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 
```
A rebase of this is necessary to merge with current.

pgwipeout reviewed (changes requested):
A rebase is necessary to merge with current.

pgwipeout commented:
> small update: I measured power before coil (12.4V) and after coli - 1.08V - seems that I have power issue not related with firmware. I found out that wifi and wps led are lighting very poor. Some element accidentally has broken - I have hope that's not SoC - I will continue investigation.

Dumb question, could the power switch had failed?

adamhnat committed:
ipq806x: Add support for Netgear Nighthawk Pro Gaming XR500
Original patch from saksev from openwrt forum
(https://forum.openwrt.org/t/netgear-xr500-under-dd-wrt-to-openwrt/30498/6
)
Original pull request Anthoni O'Brien #3052 - dropped author doesn't
respond for some time - I would like to continue - this pull includes
code fixes according to adrianschmutzler and hnyman comments on original
request
Manufacturer page: https://www.netgear.com/gaming/xr500/
Specifications:
Almost identical to Netgear R7800(https://openwrt.org/toh/netgear/r7800)
Differences:
Outside casing
Bigger flash storage
Removed esata
swapped leds: (r7800 > xr500)
usb1 (gpio 7 > 8)
usb2 (gpio 8 > 26)
guest/esata (gpio 26 > 7)
SoC: Qualcomm Atheros IPQ8065
RAM: 512 MB
Storage: 256 MiB
Wireless: Qualcomm Atheros QCA9984
Ethernet: 5x 1000/100/10
USB: 2x 3.0
Install via TFTP recovery:
1.Turn off the power, push and hold the reset button (in a hole on
backside) with a pin
2.Turn on the power and wait till power led starts flashing white (after
it first flashes orange for a while)
3.Release the reset button and tftp the factory img in binary mode. The
power led will stop flashing if you succeeded in transferring the image,
and the router reboots rather quickly with the new firmware.
4.Try to ping the router (ping 192.168.1.1). If does not respond, then
tftp will not work either.
Uploading the firmware image with a TFTP client
$ tftp 192.168.1.1
bin
put openwrt-ipq806x-netgear_xr500-squashfs-factory.img
Signed-off-by: Adam Hnat adamhnat@gmail.com
Tested-By: Peter Geis pgwipeout@gmail.com

adamhnat commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,412 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+#include "qcom-ipq8065.dtsi"
+#include <dt-bindings/input/input.h>
+
```
rebased (that was not easy)

adamhnat commented:
> > small update: I measured power before coil (12.4V) and after coli - 1.08V - seems that I have power issue not related with firmware. I found out that wifi and wps led are lighting very poor. Some element accidentally has broken - I have hope that's not SoC - I will continue investigation.
> 
> Dumb question, could the power switch have failed?

no, I've measured power switch - it's ok. In mean time I fixed my pull request to be rebased without conflicts.

adamhnat commented:
removed from target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi sections moved to general dtsi file same as in 57e029ba573bb6b4e19c90d55960bed0357c321b and 58d2bd88aefef1f3a37b401223fccae374af6c63

adamhnat committed:
Merge branch 'master' into xr500

adschm commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -30,11 +30,12 @@ nec,wg2600hp)
+	ucidef_set_led_usbport "usb1" "USB 1" "${boardname}:white:usb1" "usb1-port1" "usb2-port1"
+	ucidef_set_led_usbport "usb2" "USB 2" "${boardname}:white:usb2" "usb3-port1" "usb4-port1"
+	ucidef_set_led_switch "wan" "WAN" "${boardname}:white:wan" "switch0" "0x20"
+	ucidef_set_led_ide "esata" "eSATA" "${boardname}:white:esata"
```
boardname needs to be removed here (i.e. just keep what's already there.)

adschm commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,533 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
remove this file, 4.19 has been finally dropped (and move files-5.4 to files).

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+#include "qcom-ipq8065.dtsi"
+#include <dt-bindings/input/input.h>
```
Please add spaces between SPDX and includes.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+
```
remove empty line here.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+		compatible = "gpio-leds";
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+	};
```
I'd move that to the DTS files as well. Having it split is more complicated to read IMO.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+	led_pins: led_pins {
+		mux {
+			pins = "gpio7", "gpio8", "gpio9", "gpio22", "gpio23",
+				"gpio24","gpio26", "gpio53", "gpio64";
```
Please group using three gpios per line.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+
+	rgmii2_pins: rgmii2_pins {
+		tx {
+			pins = "gpio27", "gpio28", "gpio29", "gpio30", "gpio31", "gpio32" ;
```
please use two lines with three gpios per line. Space before ";" can be removed.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+				label = "kernel";
+				reg = <0x1480000 0x0400000>;
+			};
+
```
drop empty line here.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+				read-only;
+			};
+
+			artbak: art@1340000 {
```
Do we need the "artbak:" here?

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+};
+
+&pcie1 {
+	status = "okay";
```
Please add empty line after status.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -1,449 +1,79 @@
-			output-high;
-		};
-	};
+        model = "Netgear Nighthawk X4S R7800";
```
Indent is spaces now, but should be tabs (seems to affect the entire file).

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -1,449 +1,79 @@
+        };
+
+        leds {
+        led_power_white: power_white {
```
wrong indent

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -1,449 +1,79 @@
+
+        leds {
+        led_power_white: power_white {
+                        label = "r7800:white:power";
```
model (r7800:) can be dropped from label now, keeping just "white:power" etc.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,76 @@
+	};
+
+	leds {
+		
```
drop empty line here.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,76 @@
+	leds {
+		
+		led_power_white: power_white {
+			label = "xr500:white:power";
```
drop model "xr500:" ...

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,76 @@
+			label = "xr500:white:wps";
+			gpios = <&qcom_pinmux 24 GPIO_ACTIVE_HIGH>;
+		};
+
```
drop empty line here.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-nighthawk.dtsi:
```diff
@@ -0,0 +1,387 @@
+		nand-bus-width = <8>;
+		nand-ecc-step-size = <512>;
+
+		partitions {
```
Please use `partitions: partitions {` here and reference it directly in the DTS files then.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -1,449 +1,79 @@
-		};
-	};
+        nand@0 {
+                partitions {
```
use &partitions reference here directly.

adschm commented on target/linux/ipq806x/files-5.4/arch/arm/boot/dts/qcom-ipq8065-xr500.dts:
```diff
@@ -0,0 +1,76 @@
+
+&nand_controller {
+	nand@0 {
+		partitions {			
```
Use &partitions reference here directly.

adschm commented on target/linux/ipq806x/patches-4.19/0069-arm-boot-add-dts-files.patch:
```diff
@@ -26,6 +26,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq8064-wxr-2533dhp.dtb \
 +	qcom-ipq8065-nbg6817.dtb \
 +	qcom-ipq8065-r7800.dtb \
++	qcom-ipq8065-xr500.dtb \
```
Can be dropped, 4.19 has been removed.

adschm commented:
Apart of the requested changes, please remove the merge commit and use git rebase to squash all changes into one proper commit.

Renji-FR commented:
Hi, I would like to continue this work but I have never work/dev for OpenWrt.
@adschm If I fix all you ask, and do a `git pull --rebase` it is enough?

adschm commented:
This is obviously discontinued.

@Renji-FR There is a first time for everything. :-)

Indeed, you somehow have to get the changes here (either by rebasing or cherry-picking), then need to create your own branch and update it, tidy it up and squash it, and finally open a new PR.

I will close this one now.
